api/encoding/outbound_call.go

```diff
@@ -44,11 +44,18 @@ type OutboundCall struct {
 }
 
 // NewOutboundCall constructs a new OutboundCall with the given options.
-func NewOutboundCall(options ...CallOption) *OutboundCall {
+func NewOutboundCall(opts ...CallOption) *OutboundCall {
 	var call OutboundCall
-	for _, opt := range options {
-		opt.apply(&call)
+	var defaultOpts = &options{}
+	for _, opt := range opts {
+		opt.opt.apply(defaultOpts)
 	}
+
+	call.headers = defaultOpts.headers
+	call.responseHeaders = defaultOpts.responseHeaders
+	call.routingDelegate = defaultOpts.routingDelegate
+	call.shardKey = defaultOpts.shardKey
+	call.routingKey = defaultOpts.routingKey
```

Comment:
This is something we should do when having the option pattern. One of the issue of the solution here is if someone adds a new option, everywhere where the options are being used it will need to add this mapping.

One solution for this would be to do this:

```
type CallOption struct {
	option callOption
}

func (c CallOption) apply(o *OutboundCall) {
  c.option(o)
}

type callOption interface {
  apply(*OutboundCall)
}

type responseHeadersOption *map[string]string 

func (c responseHeadersOption) apply(*OutboundCall) {
  o.responseHeaders = c.headers
}

// ResponseHeaders specifies that headers received in response to this request
// should replace the given map.
func ResponseHeaders(h *map[string]string) CallOption {
	return responseHeadersOption(h)
}
```

and we can still have
```
var call OutboundCall
for _, opt := range options {
  opt.apply(&call)
}
```

Benefits are:
- it is transparent of what we have today
- no need of changing code where we apply the option
- reflect.DeepEqual will work

Note: this is not following exactly the guidance of the option pattern. But we are restricted with API and we can not break it.

Reply:
done